src/plugins/audioplayers/player.c

```diff
@@ -0,0 +1,452 @@
+        return;
+    }
+    audio_player_set_position(self, 0);
+    audio_player_resume(self);
+}
+
+void audio_player_pause(struct audio_player *self) {
+    GstStateChangeReturn ret = gst_element_set_state(self->playbin, GST_STATE_PAUSED);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
+        LOG_ERROR("Unable to set the pipeline to the paused state.");
+        return;
+    }
+    audio_player_on_position_update(self);  // Update to exact position when pausing
+}
+
+void audio_player_resume(struct audio_player *self) {
+    if (!self->is_initialized) {
+        return;
+    }
+    GstStateChangeReturn ret = gst_element_set_state(self->playbin, GST_STATE_PLAYING);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
+        LOG_ERROR("Unable to set the pipeline to the playing state.");
+        return;
+    }
+    audio_player_on_duration_update(self);  // Update duration when start playing, as no event is emitted elsewhere
+}
+
+void audio_player_dispose(struct audio_player *self) {
+    if (self->is_initialized) {
+        audio_player_pause(self);
+    }
+    gst_object_unref(self->bus);
+    self->bus = NULL;
+    gst_object_unref(self->source);
+    self->source = NULL;
+
+    gst_element_set_state(self->playbin, GST_STATE_NULL);
+    gst_object_unref(self->playbin);
+    self->playbin = NULL;
+
+    self->is_initialized = false;
+
+    if (self->url != NULL) {
+        free(self->url);
+        self->url = NULL;
+    }
+
+    if (self->player_id != NULL) {
+        free(self->player_id);
+        self->player_id = NULL;
+    }
+
+    if (self->channel != NULL) {
+        free(self->channel);
+        self->channel = NULL;
+    }
+}
+
+int64_t audio_player_get_position(struct audio_player *self) {
+    gint64 current = 0;
+    if (!gst_element_query_position(self->playbin, GST_FORMAT_TIME, &current)) {
+        LOG_ERROR("Could not query current position.");
+        return 0;
+    }
+    return current / 1000000;
+}
+
+int64_t audio_player_get_duration(struct audio_player *self) {
+    gint64 duration = 0;
+    if (!gst_element_query_duration(self->playbin, GST_FORMAT_TIME, &duration)) {
+        LOG_ERROR("Could not query current duration.");
+        return 0;
+    }
+    return duration / 1000000;
+}
+
+void audio_player_set_volume(struct audio_player *self, double volume) {
+    if (volume > 1) {
+        volume = 1;
+    } else if (volume < 0) {
+        volume = 0;
+    }
+    g_object_set(G_OBJECT(self->playbin), "volume", volume, NULL);
+}
+
+void audio_player_set_playback_rate(struct audio_player *self, double rate) {
+    audio_player_set_playback(self, audio_player_get_position(self), rate);
+}
+
+void audio_player_set_position(struct audio_player *self, int64_t position) {
+    if (!self->is_initialized) {
+        return;
+    }
+    audio_player_set_playback(self, position, self->playback_rate);
+}
+
+void audio_player_set_source_url(struct audio_player *self, char *url) {
+    if (self->url == NULL || strcmp(self->url, url)) {
```

Comment:
is `NULL` allowed for `url` here? if yes this may segfault

Reply:
ah okay

```suggestion
    DEBUG_ASSERT_NOT_NULL(url);
    if (self->url == NULL || strcmp(self->url, url)) {
```

technically this doesn't make a difference, the app will crash with or without the assert if NULL is passed for url, but it's good to document the preconditions in some way

Reply:
got it